assets/dns/daemonset.yaml

```diff
@@ -44,14 +44,64 @@ spec:
           timeoutSeconds: 5
           successThreshold: 1
           failureThreshold: 5
+      - name: dns-node-resolver
+        # image is set at runtime
+        imagePullPolicy: IfNotPresent
+        securityContext:
+          privileged: true
+        volumeMounts:
+        - name: hosts-file
+          mountPath: /etc/hosts
+        # env NAMESERVER and CLUSTER_DOMAIN are set at runtime
+        command:
+        - /bin/bash
+        - -c
+        - |
+          #!/bin/bash
+          set -euo pipefail
+
+          REGISTRY_SERVICE="docker-registry.default.svc"
+          OPENSHIFT_MARKER="openshift-generated-node-resolver"
+          HOSTS_FILE="/etc/hosts"
+          TEMP_FILE="/etc/hosts.tmp"
+
+          # Make a temporary file with the old hosts file's attributes.
+          cp -f --attributes-only "${HOSTS_FILE}" "${TEMP_FILE}"
+
+          while true; do
+            # Fetch registry service IP from cluster dns if present
+            ips=($(dig @"${NAMESERVER}" +short "${REGISTRY_SERVICE}.${CLUSTER_DOMAIN}"))
+
+            # Update /etc/hosts only if we get valid registry service IPs
+            # We will not update /etc/hosts when there is coredns service outage or api unavailability
+            # Stale entries could exist in /etc/hosts if the registry service is deleted
+            if [[ "$?" -eq 0 && "${#ips[@]}" -ne 0 ]]; then
+              # Build a new hosts file from /etc/hosts with our custom entries filtered out
+              grep -v "${OPENSHIFT_MARKER}" "${HOSTS_FILE}" > "${TEMP_FILE}"
+
+              # Append resolver entries for registry
+              for ip in "${ips[@]}"; do
```

Comment:
How can the service name resolve to a list of service IPs (unless it's headless or something weird)?

But I don't mind looping to be safe, but I wouldn't really expect to have more than one.